core/src/main/scala/org/apache/spark/sql/streaming/DirectKafkaStreamSource.scala

```diff
@@ -48,31 +53,112 @@ final class DirectKafkaStreamRelation(
     extends StreamBaseRelation(opts)
     with Logging with Serializable {
 
-  val topicsSet = options("topics").split(",").toSet
+  val topics = options("subscribe").split(",").toList
   val kafkaParams: Map[String, String] = options.get("kafkaParams").map { t =>
     t.split(";").map { s =>
       val a = s.split("->")
       (a(0), a(1))
     }.toMap
   }.getOrElse(Map())
+  val preferredHosts = LocationStrategies.PreferConsistent
+  val startingOffsets = JsonUtils.partitionOffsets(options("startingOffsets"))
 
-  val K = options.getOrElse("K", "java.lang.String")
-  val V = options.getOrElse("V", "java.lang.String")
-  val KD = options.getOrElse("KD", "kafka.serializer.StringDecoder")
-  val VD = options.getOrElse("VD", "kafka.serializer.StringDecoder")
+//  val K = options.getOrElse("K", "java.lang.String")
+//  val V = options.getOrElse("V", "java.lang.String")
+//  val KD = options.getOrElse("KD", "kafka.serializer.StringDecoder")
+//  val VD = options.getOrElse("VD", "kafka.serializer.StringDecoder")
 
   override protected def createRowStream(): DStream[InternalRow] = {
-    val ck: ClassTag[Any] = ClassTag(Utils.getContextOrSparkClassLoader.loadClass(K))
-    val cv: ClassTag[Any] = ClassTag(Utils.getContextOrSparkClassLoader.loadClass(V))
-    val ckd: ClassTag[Decoder[Any]] = ClassTag(Utils.getContextOrSparkClassLoader.loadClass(KD))
-    val cvd: ClassTag[Decoder[Any]] = ClassTag(Utils.getContextOrSparkClassLoader.loadClass(VD))
-    KafkaUtils.createDirectStream[Any, Any, Decoder[Any], Decoder[Any]](context,
-      kafkaParams, topicsSet)(ck, cv, ckd, cvd).mapPartitions { iter =>
-      val encoder = RowEncoder(schema)
-      // need to call copy() below since there are builders at higher layers
-      // (e.g. normal Seq.map) that store the rows and encoder reuses buffer
-      iter.flatMap(p => rowConverter.toRows(p._2).iterator.map(
-        encoder.toRow(_).copy()))
+//    val ck: ClassTag[Any] = ClassTag(Utils.getContextOrSparkClassLoader.loadClass(K))
+//    val cv: ClassTag[Any] = ClassTag(Utils.getContextOrSparkClassLoader.loadClass(V))
+//    val ckd: ClassTag[Decoder[Any]] = ClassTag(Utils.getContextOrSparkClassLoader.loadClass(KD))
+//    val cvd: ClassTag[Decoder[Any]] = ClassTag(Utils.getContextOrSparkClassLoader.loadClass(VD))
+
+    val consumerStrategies = ConsumerStrategies
+      .Subscribe[String, String](topics, kafkaParams, startingOffsets)
+
+    val stream = KafkaUtils.createDirectStream[String, String](context,
+      preferredHosts, consumerStrategies)
+      .mapPartitions { iter =>
+        val encoder = RowEncoder(schema)
+        iter.flatMap(p => {
+          rowConverter.toRows(p.value()).iterator.map(
+          encoder.toRow(_).copy())
+        })
+      }
+    stream
+  }
+}
+
+/**
+  * Copied from org.apache.spark.sql.kafka010.JsonUtils
+  */
+private object JsonUtils {
```

Comment:
JsonUtil can be a separate class. Only saw usage in test classes. If so we can mode it to test packages.